finetune/nn/graph_conv.py

```diff
@@ -0,0 +1,185 @@
+        )
+
+    def call(self, input):
+        projected = tf.matmul(input, self.kernel)
+        reshaped = tf.reshape(
+            projected, tf.concat(tf.shape(projected)[:-1], [self.num_outputs, -1])
+        )  # batch, seq, num_outputs, feats / num_outputs (new_feats)
+
+        transposed = tf.transpose(
+            a=reshaped, perm=[0, 2, 1, 3]
+        )  # batch, num_out, seq, feats
+        similarity = tf.matmul(
+            transposed, transposed, transpose_b=True
+        )  # batch, num_out, seq, seq
+        return tf.transpose(similarity, [0, 2, 3, 1])  # batch, seq, seq, num
+
+
+class GraphSmoothing(tf.keras.layers.Layer):
+    def __init__(self, num_classes, num_convolutions=1):
+        super().__init__()
+        self.num_classes = num_classes
+        self.num_convolutions = num_convolutions
+
+    def build(self, input_shape):
+        self.final_gcn = GraphConvolution(self.num_classes)
+        self.other_gcns = [
+            GraphConvolution(int(input_shape[0][-1]))
+            for _ in range(self.num_convolutions - 1)
+        ]
+
+    def call(self, inputs):
+        features, positions, sequence_lengths = inputs
+        graph = graph_heads(positions, sequence_lengths)
+        adjacency_mat = tf.stack(
+            [graph["above"], graph["below"], graph["left"], graph["right"]], 1
+        )
+        for gcn in self.other_gcns + [self.final_gcn]:
+            features = gcn([features, adjacency_mat])
+
+        return features
+
+
+class GraphConvolution(tf.keras.layers.Layer):
+    def __init__(self, num_outputs):
+        super().__init__()
+        self.num_outputs = num_outputs
+
+    def build(self, input_shape):
+        feature_shape, adjacency_mat_shape = input_shape
+        feature_dim = int(feature_shape[-1])
+        adjacency_heads = int(adjacency_mat_shape[1])
+        self.kernel = self.add_weight(
+            "kernel", shape=[adjacency_heads, feature_dim, self.num_outputs]
+        )
+        self.bias = self.add_weight("kernel", shape=[self.num_outputs])
+
+        self.identity_kernel = self.add_weight(
+            "kernel", shape=[feature_dim, self.num_outputs]
+        )
+
+    def call(self, inputs):
+        features, adjacency_mat = inputs
+        features_with_head_dim = tf.expand_dims(features, 1)  # batch, 1, seq, feats
+        # hopefully this will broadcast across heads dim?
+        projected = tf.matmul(features_with_head_dim, self.kernel)
+        # projected = batch, heads, seq, n_out
+        # adjacency_mat = batch, heads, seq, seq
+        adjacency_mat = tf.nn.relu(adjacency_mat)  # remove any spurious negative values
+        normed_adjacency_mat = tf.math.divide_no_nan(
+            adjacency_mat, tf.reduce_sum(adjacency_mat, 2, keepdims=True)
+        )
+        shared = tf.matmul(
+            projected, normed_adjacency_mat, transpose_a=True
+        )  # batch, n_out, heads, seq
+        identity_out = tf.matmul(features, self.identity_kernel)
```

Comment:
The graph where everything is connected only to its self

Reply:
I'm not sure I understand what you mean here? It seems like identity_kernel and kernel are both just learned weights, with the latter getting an additional smoothing applied based on the adjacency matrix afterwards?

Reply:
So the adjacency matrices explicitly do not connect nodes to themselves.

Reply:
To achieve that, would it make sense to be adding projected and shared? identity_out seems to be different, since it's derived from a completely different set of weights applied to features. Is there a reason for that in particular?

Reply:
So each of the graphs has it's own independent sets of weights as well they just happen to be in the same tensor.

Reply:
So unless I've messed something up this should be exactly equivalent to having an additional graph that is the identity matrix

Reply:
ah got it. I thought you wanted to add nodes to themselves in the already existing graphs, but this also seems reasonable